File: Commandline/Parser.php
#! /usr/local/bin/php
<?php

function showHelp() {
	print <<<"Help"

Usage: parser [verb] [input file path] ([output file path])

verb:
   help: show this help
   b: base64 encoding
   c: compress
   bc: both compressing and base64 encoding
   x: neither compressing nor base64 encoding


Help;
}

function omitComment($s,$st,$en) {
	$r="";
	$l=preg_split("/$en/",$s);
	foreach ($l as $p) {
		$r.=preg_split("/$st/",$p)[0];
	}
	return $r;
}

function compress($s) {
	$s=omitComment($s, "\\/\\*", "\\*\\/");
	$s=omitComment($s, "<!\\-\\-", "\\-\\->");
	$s=preg_replace("/[ \t]+\/\/.+$/m","",$s);
	$s=preg_replace("/^\/\/.+$/m","",$s);
	$s=preg_replace("/\<\/?editing\-folder\>/","",$s);
	$s=preg_replace("/ +xml:space=[\"']?preserve[\"']? +/"," ",$s);
	$s=preg_replace("/ +xml:space=[\"']?preserve[\"']? */","",$s);
	foreach (array(";",":",",","{","}") as $t) {
		$s=preg_replace("/${t}[ \t\n\v\f\r]*[\n\v\f\r][ \t\n\v\f\r]*/",$t,$s);
	}
	$s=preg_replace("/  +/"," ",$s);
	$s=preg_replace("/^ /m","",$s);
	$s=preg_replace("/ $/m","",$s);
	$s=preg_replace("/[\t\n\v\f\r]+/","",$s);
	$s=preg_replace("/> +/",">",$s);
	return $s;
}

$l=count($argv);

if ($l>1 and $argv[0]=="help") {
	showHelp();
	exit(0);
}
elseif ($l>2) {
	$b=preg_match("/b/i",$argv[1]);
	$c=preg_match("/c/i",$argv[1]);
	if ($l>3) {
		$sa=$argv[3];
	}
	else {
		$sa=null;
	}
	$r="";
	$inf=$argv[2];
	if (is_readable($inf)) {
		$m=mime_content_type($inf);
		$f=fopen($inf,"rb");
		$r=fread($f,filesize($inf));
		fclose($f);
		if ($c) {
			$r=compress($r);
		}
		if ($b) {
			$r="data:$m;base64,".base64_encode($r);
		}
	}
	else {
		fputs(STDERR,"Cannot read this file or does not exist: $inf\n");
		exit(1);
	}
	if ($sa != null) {
		if (is_writable($sa) or !file_exists($sa)) {
			$f=fopen($sa,"w");
			fwrite($f,$r);
			fclose($f);
			exit(0);
		}
		else {
			fputs(STDERR,"Cannot write into this file: $sa\n");
			exit(1);
		}
	}
	else {
		print $r;
		exit(0);
	}
}
else {
	fputs(STDERR,"Not enough arguments!\n\n");
	showHelp();
	exit(1);
}

?>

File: Commandline/Parser.py
#! /usr/local/bin/python3.7
# -*- coding: utf-8 -*-

import os
import sys
import re
import base64
import mimetypes

def showHelp():
	print("\nUsage: parser [verb] [input file path] ([output file path])\n")
	print("verb:")
	print("   help: show this help")
	print("   b: base64 encoding")
	print("   c: compress")
	print("   bc: both compressing and base64 encoding")
	print("   x: neither compressing nor base64 encoding")
	print("\n")

def exist(s,p):
	return re.search(re.compile("(?i)"+p),s) != None

def omitComment(s,st,en):
	r=""
	l=s.split(en)
	for p in l:
		r+=p.split(st)[0]
	return r

def compress(s):
	s=omitComment(s, "/*", "*/")
	s=omitComment(s, "<!--", "-->")
	s=re.sub(r"(?m)[ \t]+\/\/.+$","",s)
	s=re.sub(r"(?m)^\/\/.+$","",s)
	s=re.sub(r"\<\/?editing\-folder\>","",s)
	s=re.sub(r" +xml:space=[\"']?preserve[\"']? +"," ",s)
	s=re.sub(r" +xml:space=[\"']?preserve[\"']? *","",s)
	for t in [";",":",",","{","}"]:
		s=re.sub(re.compile("(?m)"+t+"[\\n\\t ]*\\n[\\n\\t ]*"),t,s)
	s=re.sub(r"  +"," ",s)
	s=re.sub(r"(?m)^ ","",s)
	s=re.sub(r"(?m) $","",s)
	s=re.sub(r"[\t\n]+","",s)
	s=re.sub(r"> +",">",s)
	return s

args=sys.argv
l=len(args)

if l>1 and args[1]=="help":
	showHelp()
	exit(0)
elif l>2:
	b = exist(args[1], "b")
	c = exist(args[1], "c")
	if l>3:
		sa = args[3]
	else:
		sa = None
	r=""
	inf=args[2]
	if os.path.isfile(inf):
		m=mimetypes.guess_type(inf)[0]
		if b and not c:
			try:
				f=open(inf,"rb")
			except:
				sys.stderr.write("Cannot read this file or does not exist: "+inf+"\n")
				exit(1)
			r=base64.b64encode(f.read()).decode()
			f.close()
			r="data:"+m+";base64,"+r
		else:
			try:
				f=open(inf,"r")
			except:
				sys.stderr.write("Cannot read this file or does not exist: "+inf+"\n")
				exit(1)
			r=f.read()
			f.close()
			if c:
				r=compress(r)
			if b:
				r="data:"+m+";base64,"+base64.b64encode(r.encode()).decode()
		if sa != None:
			try:
				f=open(sa,"w")
			except:
				sys.stderr.write("Cannot write into this file: "+sa+"\n")
				exit(1)
			f.write(r)
			f.close()
		else:
			sys.stdout.write(r)
		exit(0)
	else:
		sys.stderr.write("Cannot read this file or does not exist: "+inf+"\n")
		exit(1)
else:
	sys.stderr.write("Not enough arguments!\n\n")
	showHelp()
	exit(1)
